classpath/java/net/Socket.java

```diff
@@ -93,7 +93,9 @@ public int read(byte[] buffer) throws IOException {
 				size = recv(sock, buffer, index, Math.min(fullSize, Socket.BUFFER_SIZE));
 				fullSize -= size;
 				index += size;
-			} while (fullSize != 0 && size != 0);
+			} while (fullSize != 0 && size > 0);
+			if(size < 0) throw new IOException("Error whilereading stream");
+			if(index == 0 && size == 0) return -1;
```

Comment:
I would also like to point out that there is a very common pattern where you can run into zero-length reads, if you're using a length-delimited protocol (say, protobuf, for instance):

* Read the length off the wire (say, of a string you're trying to read)
* Allocate a byte array of the appropriate size
* Pass that to `read(...)` to read the appropriate number of bytes.

If you don't gracefully support zero-length reads, that's a bug waiting to happen.  I ran into that exact issue not long ago with a Rust decompression library.

From a security perspective, that's a very dangerous bug to have, since that's an edge case that not many people would have considered (when writing the code).  All sorts of weird stuff could fall out of that.